### PR TITLE
rgw: fix the signature mismatch of FormPost in swift API

### PR DESCRIPTION
**Pull request by Jing-Scott**
Fixes: http://tracker.ceph.com/issues/20220
Signed-off-by: Jing Wenjun <jingwenjun@cmss.chinamobile.com>

### CODE REVIEW

**Comment by Jing-Scott** [comment]
The blow are the test steps, which also can be found [here](https://docs.openstack.org/developer/swift/api/form_post_middleware.html#form-post-format).

- we can use the python code to generate `Expires` and `Signature` in ceph.
```python
 import hmac
 from hashlib import sha1
 from time import time
 path = '/swift/v1/form-1/'
 redirect = ''
 max_file_size = 5373952000
 max_file_count = 2
 expires = int(time() + 30000)
 key = 'mykey'
 hmac_body = '%s\n%s\n%s\n%s\n%s' % (path, redirect,
                 max_file_size, max_file_count, expires)
 signature = hmac.new(key, hmac_body, sha1).hexdigest()
 print 'Expires: ' + str(expires)
 print 'Signature: ' + signature
```

- then copy the values of `Expires` and `Signature` to the blow xml script.
```
<form action="http://<you_domain>/swift/v1/form-1/" method="POST" enctype="multipart/form-data">
  <input type="hidden" name="max_file_size" value="5373952000" />
  <input type="hidden" name="max_file_count" value="2" />
  <input type="hidden" name="expires" value="1496934665" />
  <input type="hidden" name="signature" value="c53d96db1b341db8b76168e487d956d307ea7cb5" />
  <!-- This signature allows for at most 2 files, -->
  <!-- but it may also have any smaller number. -->
  <!-- Remove file inputs as needed. -->
  <input type="file" name="file0" />
  <br />
  <input type="file" name="file1" />
  <br />
  <input type="submit" />
</form>
```

- At last, we can open the upper xml directly in your browser to upload objects.

**Comment by Jing-Scott** [comment]
@rzarzynski hi, how about this fix? thanks!

File: src/rgw/rgw_rest_swift.cc [hunk 1]
```diff
@@ -1943,9 +2003,9 @@ bool RGWFormPost::is_next_file_to_upload()
 
     const auto field_iter = part.fields.find("Content-Disposition");
     if (std::end(part.fields) != field_iter) {
-      const auto& params = field_iter->second.params;
+      auto& params = field_iter->second.params;
 
-      if (std::end(params) != params.find("filename")) {
+      if (std::end(params) != params.find("filename") && ! params["filename"].empty()) {
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_swift.cc]: The `operator[]` can unnecessarily modify the `params`. Maybe replace with something like that:
```cpp
      const auto& params = field_iter->second.params;
      const auto& filename_iter = params.find("filename");
      if (std::end(params) != filename_iter && ! filename_iter->empty()) {
```

Comment (Jing-Scott) [inline src/rgw/rgw_rest_swift.cc]: @rzarzynski done!

Comment (Jing-Scott) [inline src/rgw/rgw_rest_swift.cc]: change from `filename_iter->empty()` to `filename_iter->second.empty()`

**Comment by rzarzynski** [comment]
jenkins retest this please

**Comment by rzarzynski** [comment]
jenkins retest this please (`Agent went offline during the build`)

**Comment by rzarzynski** [comment]
Shaman [is fine](https://shaman.ceph.com/builds/ceph/wip-rgw-auth-rework-cont-3-awsv4/33768cd13e1c8329957e1f9ac4bd03ee1c10cd05/) with the changes.

**Comment by tchaikov** [comment]
retest this please.

**Comment by Jing-Scott** [comment]
jenkins retest this please

**Comment by Jing-Scott** [comment]
@rzarzynski thanks for your review again!

**Pull request merged**